Wrappers/Python/cil/optimisation/algorithms/SIRT.py

```diff
@@ -17,38 +17,96 @@
+                or :class:`.TotalVariation` function and :meth:`.TotalVariation.proximal`.
+
+    kwargs:
+        Keyword arguments used from the base class :class:`.Algorithm`.    
+
+    Note 
+    ----
+    
+    If :code:`constraint` is not passed, then :code:`lower` and :code:`upper` are looked at and an :class:`.IndicatorBox`
+    function is created.
+
+    If :code:`constraint` is passed, :code:`proximal` method is required to be implemented.
+
+    Note
+    ----
+
+    The preconditioning arrays (weights) :code:`M` and :code:`D` used in SIRT are defined as
+
+    .. math:: M = \frac{1}{A*\mathbb{1}} = \frac{1}{\sum_{j}a_{i,j}}
+
+    .. math:: D = \frac{1}{A*\mathbb{1}} = \frac{1}{\sum_{i}a_{i,j}}
+
+    In case of division errors above, :meth:`._fix_weights` can be used, where :code:`np.nan`, :code:`+np.inf` and :code:`-np.inf` values
+    are replaced with 1.0.
+
 
-class SIRT(Algorithm):
 
-    r'''Simultaneous Iterative Reconstruction Technique
     
-    Problem: 
+
     
-    .. math::  
     
-      A x = b
+    Examples
+    --------
+    .. math:: \underset{x}{\mathrm{argmin}} \| x - 5\|^{2}
     
-    :param initial: Initial guess
-    :param operator: Linear operator for the inverse problem
-    :param data: Acquired data to reconstruct       
-    :param constraint: Function proximal method
-                e.g.  :math:`x\in[0, 1]`, :code:`IndicatorBox` to enforce box constraints
-                        Default is :code:`None`).
-    '''
-    def __init__(self, initial=None, operator=None, data=None, lower=None, upper=None, constraint=None, **kwargs):
-        '''SIRT algorithm creator
-
-       Optional parameters:
-
-      :param initial: Initial guess
-      :param operator: Linear operator for the inverse problem
-      :param data: Acquired data to reconstruct 
-      :param lower: Scalar specifying lower bound constraint on pixel values, default -inf
-      :param upper: Scalar specifying upper bound constraint on pixel values, default +inf
-      :param constraint: More general constraint, given as Function proximal method
-                   e.g.  :math:`x\in[0, 1]`, :code:`IndicatorBox` to enforce box constraints
-                         Default is :code:`None`). constraint takes priority over lower and upper.'''
+    >>> from cil.framework import ImageGeometry
+    >>> from cil.optimisation.operators import IdentityOperator    
+    >>> from cil.optimisation.algorithms import SIRT
+    >>> ig = ImageGeometry(3,4)
+    >>> A = IdentityOperator(ig)
+    >>> b = ig.allocate(5.0)
+    >>> sirt = SIRT(initial = ig.allocate(0), operator = A, data=b, max_iteration=5) 
+    SIRT setting up
+    SIRT configured
+    >>> sirt.run(verbose=0)
+    >>> sirt.solution.array
+    array([[5., 5., 5.],
+           [5., 5., 5.],
+           [5., 5., 5.],
+           [5., 5., 5.]], dtype=float32)
+    """    
+
+
+    def __init__(self, initial, operator, data, lower=None, upper=None, constraint=None, **kwargs):
```

Comment:
do we need `lower` `upper` and `constraint` when `constraint could be an `IndicatorBoxFunction`?

Reply:
In practice, we do not need: `lower`, `upper` and `constraint`. `constraint` acts as a `Function` and inside the `update` method we use its `proximal` method. 

We can have a `constraint` which is defined by an `IndicatorBox` (which has a `proximal` method), or create an `IndicatorBox` using `lower` and `upper`. In practice, since `SIRT` behaves exactly as `ISTA` and `FISTA`, `constraint` should be a `Function` with a `proximal`, e.g., `TotalVariation`, `FGP_TV`, `L1`, `L2NormSquared` etc.

However, changing the signature of SIRT is a big change and in most of our demos, papers we are using lower, upper and constraint.